### DISCUSSION TITLE
Installing LegacyUpdate on Windows XP using only the command line

### DISCUSSION THREAD

**Discussion by ohault** [LegacyUpdate/LegacyUpdate#294] [Questions & Help]
From a freshly installed Windows XP or Windows Server 2003, how to download and install LegacyUpdate using only the command line prompt?

**Comment by kirb** [comment]
In XP/2003, there’s no straightforward way to download a file from the command line that I know of. The closest solution is to write up a VBScript like [this one](https://serverfault.com/a/29720).

That said, you can always download the latest version by going directly to `http://legacyupdate.net/LegacyUpdate.exe` in Internet Explorer.

**Reply by ohault** [reply]
Maintaining http://legacyupdate.net/LegacyUpdate.exe is I guess a very good start.

I'm convinced there should be a way with a single command "to paste" into the command prompt to get it done.

I'm currently trying to explore different avenues to find a definitive answer to this question.

- FTP access to LegacyUpdate.exe
- TELNET on port 80
- Compacting a VB and/or JavaScript
- uuencoding/decoding
- certutil (with Windows Server 2003 tools)
- bitsadmin (from a cab file in the iso /support/tools/ )
- ....

This would require to use with piping, std input and output redirection from command line and/or & to fit alls into a single prompt.

Help is welcomed in this quest.

**Reply by kirb** [reply]
Not entirely sure I’d support protocols like FTP/telnet, as the site is behind Cloudflare, which doesn’t really like non-HTTP protocols (this is in their ToS). I’d need some more convincing to expose an IP on a subdomain for this 😉

Maybe I could support WebDAV, but of course that’s accessed by UNC paths, which cmd doesn’t support. At that point, you can only type it into Run, an Explorer/IE address bar, or via argument to `explorer`, and it’s no easier/different to type `\\legacyupdate.net\LegacyUpdate.exe` as compared to `http://legacyupdate.net/LegacyUpdate.exe` (actually WebDAV would be slower because it’ll attempt SMB first and time out). WebDAV would be neat to allow for browsing through a file hierarchy, though.

The remaining options would be neat, but they need extra prep than just clicking the big button on legacyupdate.net or directly typing in the LegacyUpdate.exe url. Please let me know if there’s a benefit I’m missing though.

**Reply by ohault** [reply]
Thank you, indeed WebDAV could be a very good avenue too. Could you test it ?

The only goal here is to find a solution to install Legacy Update without any user intereactions from a freshly installed Windows, .

BITSADMIN (not included in Windows XP Home edition) 

From a freshly Windows Server 2003 with support tools installed from the ISO /support/tools/support.msi, and admin rights, I'm currently blocked with the following. 

`C:\>bitsadmin /create MyJob

BITSADMIN version 1.5 [ 5.2.3790.0 ]
BITS administration utility.
(C) Copyright 2000-2002 Microsoft Corp.

Created job {108AF9A3-BF26-476F-B760-2586D1111E9D}.

C:\>bitsadmin /addfile {108AF9A3-BF26-476F-B760-2586D1111E9D} http://legacyupdate.net/LegacyUpdate.exe c:\LegacyUpdate.exe

BITSADMIN version 1.5 [ 5.2.3790.0 ]
BITS administration utility.
(C) Copyright 2000-2002 Microsoft Corp.

Unable to add file to job - 0x800704dd
The operation being requested was not performed because the user has not logged on to the network.
The specified service does not exist.
`

Another try from another VM, command launched from the Windows Support Tools Command Prompt in Windows Server 2003:
`bitsadmin /create MyJob & bitsadmin /addfile MyJob http://legacyupdate.net/LegacyUpdate.exe c:\LegacyUpdate.exe & pause & bitsadmin /resume MyJob & pause & bitsadmin /complete MyJob & pause & call c:\LegacyUpdate.exe
`

FTP could be automated with a -s:commands.txt file, some FTP gateway could be considered.  Here is an example of a commercial one https://sftpcloud.io/ftp-and-sftp-using-cloudflare-r2

TELNET is more complicated but similar to old days BBS, I'm quite sure downloading a file could be possible, and perhaps a binalry one.

CERTUTIL looks like to work, but would require the download file to be encoded (base64 ?) with "certutil.exe -decode LegacyUpdate.txt LegacyUpdate.exe" 
`C:\>certutil -split -f -URLcache http://legacyupdate.net/LegacyUpdate.txt
****  Online  ****
WinHttp Cache entries: 0

CertUtil: -URLCache command completed successfully.`

Inline Windows VBScript using WHS
Thanks and credits to : 
- https://stackoverflow.com/users/397667/bw-%c3%bcezi and https://stackoverflow.com/users/246342/alex-k for VBS download script https://stackoverflow.com/questions/2973136/download-a-file-with-vbs
- [5422m4n](https://stackoverflow.com/users/1252968/5422m4n), [MultiplyByZer0] for oneline vbs https://stackoverflow.com/questions/77402797/how-to-convert-a-vbscript-into-a-one-liner
-  (https://stackoverflow.com/users/2688027/multiplybyzer0), [Martin](https://stackoverflow.com/users/770/martin)  for timeout https://stackoverflow.com/questions/1672338/how-to-sleep-for-five-seconds-in-a-batch-file-cmd, 
- [PA.](https://stackoverflow.com/users/30447/pa) for Inline vbscript https://stackoverflow.com/questions/13115508/theres-any-way-to-run-vbscript-directly-to-wscript-or-cscript-command-line 

On my Windows XP test system, pasting the following command line will launch LegacyUpdate installer
`echo dim xHttp: Set xHttp = createobject("Microsoft.XMLHTTP"):dim bStrm: Set bStrm = createobject("Adodb.Stream"):xHttp.Open "GET", "http://legacyupdate.net/LegacyUpdate.exe", False:xHttp.Send::with bStrm: .type = 1 : .open: .write xHttp.responseBody: .savetofile "LegacyUpdate.exe", 2 :end with > temp.vbs && ping 127.0.0.1 -n 6 > nil && wscript temp.vbs && ping 127.0.0.1 -n 6 > nil && del temp.vbs && Call LegacyUpdate.exe`

From Windows Server 2003 RTM, the http://legacyupdate.net/LegacyUpdate.exe redirect would not be supported =>
`echo dim xHttp: Set xHttp = createobject("Microsoft.XMLHTTP"):dim bStrm: Set bStrm = createobject("Adodb.Stream"):xHttp.Open "GET", "http://content.legacyupdate.net/legacyupdate/LegacyUpdate-1.10.exe", False:xHttp.Send::with bStrm: .type = 1 : .open: .write xHttp.responseBody: .savetofile "LegacyUpdate.exe", 2 :end with > temp.vbs && ping 127.0.0.1 -n 6 > nil && wscript temp.vbs && ping 127.0.0.1 -n 6 > nil && del temp.vbs && Call LegacyUpdate.exe
`

It's yet not far to be perfect, but it illustrates the concept. I don't understnad with I should add two timeouts to wait enough time meanwhile && is supposed to run sequencialy. 

It should be also possible to avoid completly using a temp file like temp.vbs (https://stackoverflow.com/questions/9074476/is-it-possible-to-embed-and-execute-vbscript-within-a-batch-file-without-using-a)

If LegacyUpdate.exe could support some command line parameters, I would be possible to further automate.

**Reply by kirb** [reply]
Thanks so much for being so detailed!

I set up WebDAV at `\\content.legacyupdate.net\content`. Unfortunately I found that it’s not just initially slow - it’s *incredibly* slow as a whole. Windows’s WebDAV implementation is kinda just terrible… I’m sure it’s better than nothing for some use cases though. It’s perfectly fast when I connect via Cyberduck or WinSCP, so it’s not the server or the protocol being slow, it’s definitely Windows. Even macOS Finder, which is usually impressively slow with mounted network shares, handles it fine.

I super strongly recommend using it via `\\content.legacyupdate.net@SSL\content` (note `@SSL`), because that forces Windows to immediately recognise it as WebDAV, and not spend a bunch of time trying to do SMB and timing out. But that’s not going to be compatible with 2000/XP… oh well. (edit: The doc I link below actually says `@SSL` wasn’t implemented till Vista anyway)

We do have command line flags - see https://legacyupdate.net/faq/automation. In this case you want `/S` (case sensitive). In 1.10.1 I’ll be adding `/s` (lowercase) as an alias, and `/passive`, which will have the same effect but still displays the UI for progress.

**Reply by kirb** [reply]
I totally didn’t consider WebDAV can be mounted with `net use`. So now we have a solution:

```batch
net use L: \\content.legacyupdate.net\content
L:\legacyupdate\LegacyUpdate.exe /S
```

I’d see about coming up with a shorter url than this, but adding redirects into the mix of an already very slow experience seems like a bad idea.

There are also some [Microsoft WebDAV extensions](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-wdv/bfde1057-4214-4ca5-a431-fab36ff625bc) that might have some ways to improve performance that I’ll look into. Because it would be useful in general to have WebDAV as a feature we recommend for everyone, so they can browse the other things we mirror on this domain.

**Reply by ohault** [reply]
It is currently just a proof of concept to illustrate some methods to support single inline command concept for downloading and launch a setup.

This has yet to be tested, secured and improved.
